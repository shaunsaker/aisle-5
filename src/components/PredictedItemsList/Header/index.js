import React from 'react';
import PropTypes from 'prop-types';
import { View, Text } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';

import styleConstants from '../../../styleConstants';
import styles from './styles';

import LinkText from '../../LinkText';

const propTypes = {
  headerItems: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string,
      flex: PropTypes.number,
      fieldName: PropTypes.string,
    }),
  ).isRequired,
  handleHeaderItemPress: PropTypes.func.isRequired,
  height: PropTypes.number,
  headerTestIDPrefix: PropTypes.string,
};

const defaultProps = {};

const Header = ({ headerItems, handleHeaderItemPress, height, headerTestIDPrefix }) => {
  return (
    <View style={[styles.container, { height }]}>
      {headerItems.map((item, index) => {
        const isNotLastItem = index !== headerItems.length - 1;

        return (
          <View
            key={item.text}
            style={[
              styles.row,
              { flex: item.flex },
              { marginRight: isNotLastItem ? styleConstants.dimensions.spacing.horizontal : 0 },
            ]}
          >
            <LinkText
              text={item.text}
              handlePress={() => handleHeaderItemPress(item.fieldName)}
              testID={`${headerTestIDPrefix}${item.text}`}
            />

            <Icon name="swap-vert" style={styles.icon} />
          </View>
        );
      })}
    </View>
  );
};

Header.propTypes = propTypes;
Header.defaultProps = defaultProps;

export default Header;
