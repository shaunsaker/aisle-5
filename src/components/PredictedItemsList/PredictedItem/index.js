import React from 'react';
import PropTypes from 'prop-types';
import { View, Text } from 'react-native';

import styles from './styles';

import ProgressBar from './ProgressBar';

const propTypes = {
  name: PropTypes.string,
  remainingQuantityPercentage: PropTypes.number,
  daysLeftUntilZeroQuantity: PropTypes.number,
  height: PropTypes.number,
};

const defaultProps = {};

const PROGRESS_BAR_WIDTH = 160;

const PredictedItem = ({
  name,
  remainingQuantityPercentage,
  daysLeftUntilZeroQuantity,
  height,
}) => {
  const daysLeftText = `${Math.floor(daysLeftUntilZeroQuantity)} days`;

  return (
    <View style={[styles.container, { height }]}>
      <Text numberOfLines={1} style={styles.text}>
        {name}
      </Text>

      <View style={styles.consumptionContainer}>
        <ProgressBar value={remainingQuantityPercentage} />

        <View style={styles.daysLeftTextContainer}>
          <Text style={styles.daysLeftText}>{daysLeftText}</Text>
        </View>
      </View>
    </View>
  );
};

PredictedItem.propTypes = propTypes;
PredictedItem.defaultProps = defaultProps;

export default PredictedItem;
