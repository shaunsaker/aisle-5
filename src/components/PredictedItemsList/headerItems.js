const headerItems = [
  {
    text: 'NAME',
    flex: 1,
    fieldName: 'name',
  },
  {
    text: 'CONSUMPTION',
    flex: 2,
    fieldName: 'remainingQuantity', // this can be less than zero
  },
];

export default headerItems;
