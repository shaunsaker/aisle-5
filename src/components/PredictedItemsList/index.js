import React from 'react';
import PropTypes from 'prop-types';
import { FlatList, View } from 'react-native';

import styles from './styles';

import Header from './Header';
import PredictedItem from './PredictedItem';
import ItemSeparator from '../ItemSeparator';

import HEADER_ITEMS from './headerItems';

export default class PredictedItemsList extends React.Component {
  constructor(props) {
    super(props);

    this.renderListHeader = this.renderListHeader.bind(this);
    this.renderItem = this.renderItem.bind(this);
    this.renderItemSeparator = this.renderItemSeparator.bind(this);

    this.itemHeight = 51;

    this.state = {};
  }

  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
    handleHeaderItemPress: PropTypes.func,
    headerTestIDPrefix: PropTypes.string,
  };

  static defaultProps = {};

  renderListHeader() {
    const { handleHeaderItemPress, headerTestIDPrefix } = this.props;

    return (
      <Header
        headerItems={HEADER_ITEMS}
        handleHeaderItemPress={handleHeaderItemPress}
        height={this.itemHeight}
        headerTestIDPrefix={headerTestIDPrefix}
      />
    );
  }

  renderItem({ item }) {
    return (
      <View style={styles.itemContainer}>
        <PredictedItem {...item} height={this.itemHeight} />
      </View>
    );
  }

  renderItemSeparator() {
    return <ItemSeparator />;
  }

  render() {
    const { data } = this.props;

    return (
      <FlatList
        ref={(c) => {
          this.flatList = c;
        }}
        keyExtractor={({ id }) => id}
        data={data}
        style={styles.container}
        contentContainerStyle={styles.contentContainer}
        getItemLayout={(data, index) => ({
          length: this.itemHeight,
          offset: this.itemHeight * index,
          index,
        })}
        bounces={false}
        ListHeaderComponent={this.renderListHeader}
        renderItem={this.renderItem}
        ItemSeparatorComponent={this.renderItemSeparator}
      />
    );
  }
}
