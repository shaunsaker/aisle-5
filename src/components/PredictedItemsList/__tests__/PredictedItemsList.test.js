import React from 'react';
import renderer from 'react-test-renderer';

import PredictedItemsList from '..';
import PREDICTED_ITEMS_LIST from '../../../mockData/predictedItemsList';

jest.mock('../Header', () => 'Header');
jest.mock('../PredictedItem', () => 'PredictedItem');
jest.mock('../../ItemSeparator', () => 'ItemSeparator');

describe('PredictedItemsList', () => {
  describe('renders', () => {
    it('renders with minimum required props', () => {
      const component = renderer.create(<PredictedItemsList data={PREDICTED_ITEMS_LIST} />);

      expect(component).toMatchSnapshot();
    });
  });
});
