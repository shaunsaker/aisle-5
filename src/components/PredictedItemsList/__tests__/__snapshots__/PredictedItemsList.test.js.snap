// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredictedItemsList renders renders with minimum required props 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  ListHeaderComponent={[Function]}
  contentContainerStyle={Object {}}
  data={
    Array [
      Object {
        "averageUsagePerDay": 0.5,
        "daysLeftUntilZeroQuantity": null,
        "id": "-1",
        "name": "Whiskey",
        "previouslyPurchasedQuantity": null,
        "remainingQuantity": null,
        "remainingQuantityPercentage": null,
      },
      Object {
        "averageUsagePerDay": 0.5,
        "daysLeftUntilZeroQuantity": 0,
        "id": "0",
        "name": "Wine",
        "previouslyPurchasedQuantity": 1,
        "remainingQuantity": 0,
        "remainingQuantityPercentage": 0,
      },
      Object {
        "averageUsagePerDay": 2,
        "daysLeftUntilZeroQuantity": 2,
        "id": "1",
        "name": "Beer",
        "previouslyPurchasedQuantity": 6,
        "remainingQuantity": 2,
        "remainingQuantityPercentage": 25,
      },
      Object {
        "averageUsagePerDay": 0.5,
        "daysLeftUntilZeroQuantity": 2,
        "id": "10",
        "name": "Bread",
        "previouslyPurchasedQuantity": 2,
        "remainingQuantity": 1,
        "remainingQuantityPercentage": 50,
      },
      Object {
        "averageUsagePerDay": 0.25,
        "daysLeftUntilZeroQuantity": 4,
        "id": "2",
        "name": "Milk",
        "previouslyPurchasedQuantity": 4,
        "remainingQuantity": 2,
        "remainingQuantityPercentage": 75,
      },
      Object {
        "averageUsagePerDay": 0.1,
        "daysLeftUntilZeroQuantity": 6,
        "id": "3",
        "name": "Cheese",
        "previouslyPurchasedQuantity": 1,
        "remainingQuantity": 1,
        "remainingQuantityPercentage": 100,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={
    Array [
      0,
    ]
  }
  style={Object {}}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Header
        headerItems={
          Array [
            Object {
              "fieldName": "name",
              "flex": 1,
              "text": "NAME",
            },
            Object {
              "fieldName": "remainingQuantity",
              "flex": 2,
              "text": "CONSUMPTION",
            },
          ]
        }
        height={51}
      />
    </View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <PredictedItem
          averageUsagePerDay={0.5}
          daysLeftUntilZeroQuantity={null}
          height={51}
          id="-1"
          name="Whiskey"
          previouslyPurchasedQuantity={null}
          remainingQuantity={null}
          remainingQuantityPercentage={null}
        />
      </View>
      <ItemSeparator />
    </View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <PredictedItem
          averageUsagePerDay={0.5}
          daysLeftUntilZeroQuantity={0}
          height={51}
          id="0"
          name="Wine"
          previouslyPurchasedQuantity={1}
          remainingQuantity={0}
          remainingQuantityPercentage={0}
        />
      </View>
      <ItemSeparator />
    </View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <PredictedItem
          averageUsagePerDay={2}
          daysLeftUntilZeroQuantity={2}
          height={51}
          id="1"
          name="Beer"
          previouslyPurchasedQuantity={6}
          remainingQuantity={2}
          remainingQuantityPercentage={25}
        />
      </View>
      <ItemSeparator />
    </View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <PredictedItem
          averageUsagePerDay={0.5}
          daysLeftUntilZeroQuantity={2}
          height={51}
          id="10"
          name="Bread"
          previouslyPurchasedQuantity={2}
          remainingQuantity={1}
          remainingQuantityPercentage={50}
        />
      </View>
      <ItemSeparator />
    </View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <PredictedItem
          averageUsagePerDay={0.25}
          daysLeftUntilZeroQuantity={4}
          height={51}
          id="2"
          name="Milk"
          previouslyPurchasedQuantity={4}
          remainingQuantity={2}
          remainingQuantityPercentage={75}
        />
      </View>
      <ItemSeparator />
    </View>
    <View
      style={null}
    >
      <View
        style={Object {}}
      >
        <PredictedItem
          averageUsagePerDay={0.1}
          daysLeftUntilZeroQuantity={6}
          height={51}
          id="3"
          name="Cheese"
          previouslyPurchasedQuantity={1}
          remainingQuantity={1}
          remainingQuantityPercentage={100}
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
